File: hmc/php/orchestrator/State.php
<?php

/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * State of both Services (HDFS, MapReduce etc.) and
 * ServiceComponents (NameNode, JobTracker etc.).
 */
class State {
  const UNKNOWN = 0;
  const UNINSTALLING = 1;
  const UNINSTALLED = 2;
  const INSTALLING = 3;
  const INSTALLED = 4;
  const STARTING = 5;
  const STARTED = 6;
  const STOPPING = 7;
  const STOPPED = 8;
  const FAILED = 9;

  public static $STATE =
    array (
      "UNKNOWN",
      "UNINSTALLING",
      "UNINSTALLED",
      "INSTALLING",
      "INSTALLED",
      "STARTING",
      "STARTED",
      "STOPPING",
      "STOPPED",
      "FAILED"
    );

  public static $DESIRED_STATE =
    array (
      self::UNINSTALLING => "uninstalled",
      self::UNINSTALLED => "uninstalled",
      self::INSTALLING => "installed_and_configured",
      self::INSTALLED => "installed_and_configured",
      self::STARTING => "running",
      self::STARTED => "running",
      self::STOPPING => "stopped",
      self::STOPPED => "stopped"
  );

  public static function getStateFromString($stateStr) {
    $found = array_keys(self::$STATE, $stateStr);
    if (is_array($found) && count($found) == 1) {
      return $found[0];
    }
    return FALSE;
  }

  /**
   * Converts INSTALLED to INSTALLING and likewise.
   * No-op for failed or unknown states
   * @param int $state
   */
  public static function convertStateToProgressingState($state) {
    $retState = self::UNKNOWN;
    switch ($state) {
      case State::UNKNOWN:
      case State::UNINSTALLING:
      case State::INSTALLING:
      case State::STARTING:
      case State::STOPPING:
      case State::FAILED:
        $retState = $state;
        break;
      case State::UNINSTALLED:
        $retState = self::UNINSTALLING;
        break;
      case State::INSTALLED:
        $retState = self::INSTALLING;
        break;
      case State::STARTED:
        $retState = self::STARTING;
        break;
      case State::STOPPED:
        $retState = self::STOPPING;
        break;
      default:
        $retState = self::UNKNOWN;
        break;
    }
    return $retState;
  }

}

class TransactionProgress {
  const PENDING = 0;
  const IN_PROGRESS = 1;
  const COMPLETED = 2;
  const FAILED = 3;

  public static $PROGRESS =
  array (
    "PENDING",
    "IN_PROGRESS",
    "COMPLETED",
    "FAILED"
    );

  public static function getProgressFromString($progressStr) {
    $found = array_keys(self::$PROGRESS, $progressStr);
    if (is_array($found) && count($found) == 1) {
      return $found[0];
    }
    return FALSE;
  }

}

function getTransactionProgressFromState($state) {
  $txnState = TransactionProgress::PENDING;
  switch ($state) {
    case State::UNKNOWN:
      $txnState = TransactionProgress::PENDING;
      break;
    case State::UNINSTALLING:
    case State::INSTALLING:
    case State::STARTING:
    case State::STOPPING:
      $txnState = TransactionProgress::IN_PROGRESS;
      break;
    case State::UNINSTALLED:
    case State::INSTALLED:
    case State::STARTED:
    case State::STOPPED:
      $txnState = TransactionProgress::COMPLETED;
      break;
    case State::FAILED:
      $txnState = TransactionProgress::FAILED;
      break;
    default:
      $txnState = TransactionProgress::PENDING;
  }
  return $txnState;
}

function getActionDescription($name, $action, $state) {
  /*
  $stateInfo = "";
  if ($state == "PENDING") {
    $stateInfo = "is $state";
  } else if ($state == "IN_PROGRESS") {
    $stateInfo = "is IN PROGRESS";
  } else if ($state == "COMPLETED" || $state =="FAILED") {
    $stateInfo = "has $state";
  }
  $description = $name . " " . strtolower($action) . " " . strtolower($stateInfo);
  */
  $description = $name.' '.$action;
  return $description;
}
?>

File: ambari-server/src/main/java/org/apache/ambari/server/state/State.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.ambari.server.state;

public enum State {
  /**
   * Initial/Clean state.
   */
  INIT(0),
  /**
   * In the process of installing.
   */
  INSTALLING(1),
  /**
   * Install failed.
   */
  INSTALL_FAILED(2),
  /**
   * State when install completed successfully.
   */
  INSTALLED(3),
  /**
   * In the process of starting.
   */
  STARTING(4),
  /**
   * State when start completed successfully.
   */
  STARTED(5),
  /**
   * In the process of stopping.
   */
  STOPPING(6),
  /**
   * In the process of uninstalling.
   */
  UNINSTALLING(7),
  /**
   * State when uninstall completed successfully.
   */
  UNINSTALLED(8),
  /**
   * In the process of wiping out the install.
   */
  WIPING_OUT(9),
  /**
   * In the process of upgrading the deployed bits.
   */
  UPGRADING(10),
  /**
   * Disabled master's backup state
   */
  MAINTENANCE(11),
  /**
   * State could not be determined.
   */
  UNKNOWN(12);

  private final int state;

  private State(int state) {
    this.state = state;
  }

  /**
   * Indicates whether or not it is a valid desired state.
   *
   * @return true if this is a valid desired state.
   */
  public boolean isValidDesiredState() {
    switch (State.values()[this.state]) {
      case INIT:
      case INSTALLED:
      case STARTED:
      case UNINSTALLED:
      case MAINTENANCE:
        return true;
      default:
        return false;
    }
  }

  /**
   * Indicates whether or not its a state indicating a task in progress.
   *
   * @return true if this is a state indicating progress.
   */
  public boolean isInProgressState() {
    switch (State.values()[this.state]) {
      case INSTALLING:
      case STARTING:
      case STOPPING:
      case UNINSTALLING:
      case WIPING_OUT:
      case UPGRADING:
        return true;
      default:
        return false;
    }
  }

  /**
   * Indicates whether or not it is a valid state for the client component.
   *
   * @return true if this is a valid state for a client component.
   */
  public boolean isValidClientComponentState() {
    switch (State.values()[this.state]) {
      case STARTING:
      case STARTED:
      case STOPPING:
        return false;
      default:
        return true;
    }
  }

  /**
   * Indicates whether or not the resource with this state can be removed.
   *
   * @return true if this is a removable state
   */
  public boolean isRemovableState() {
    switch (State.values()[this.state]) {
      case INIT:
      case INSTALLING:
      case INSTALLED:
      case INSTALL_FAILED:
      case UNINSTALLED:
      case UNKNOWN:
      case MAINTENANCE:
        return true;
      default:
        return false;
    }
  }
}
